numpy/_core/src/npysort/x86_simd_qsort.dispatch.cpp

```diff
@@ -0,0 +1,87 @@
+#include "x86_simd_qsort.hpp"
+#ifndef __CYGWIN__
+
+#if defined(NPY_HAVE_AVX512_SKX)
+    #include "x86-simd-sort/src/avx512-32bit-qsort.hpp"
+    #include "x86-simd-sort/src/avx512-64bit-qsort.hpp"
+    #include "x86-simd-sort/src/avx512-64bit-argsort.hpp"
+#elif defined(NPY_HAVE_AVX2)
+    #include "x86-simd-sort/src/avx2-32bit-qsort.hpp"
+    #include "x86-simd-sort/src/avx2-64bit-qsort.hpp"
+#endif
+
+template<typename T>
+void x86_qsort(T* arr, npy_intp num)
+{
+#if defined(NPY_HAVE_AVX512_SKX)
+    avx512_qsort(arr, num, true);
+#elif defined(NPY_HAVE_AVX2)
+    avx2_qsort(arr, num, true);
+#endif
+}
+
+template<typename T>
+void x86_qselect(T* arr, npy_intp num, npy_intp kth)
+{
+#if defined(NPY_HAVE_AVX512_SKX)
+    avx512_qselect(arr, kth, num, true);
+#elif defined(NPY_HAVE_AVX2)
+    avx2_qselect(arr, kth, num, true);
+#endif
+}
```

Comment:
```suggestion
namespace {
template<typename T>
inline void x86_qsort(T* arr, npy_intp num)
{
#if defined(NPY_HAVE_AVX512_SKX)
    avx512_qsort(arr, num, true);
#elif defined(NPY_HAVE_AVX2)
    avx2_qsort(arr, num, true);
#endif
}

template<typename T>
inline void x86_qselect(T* arr, npy_intp num, npy_intp kth)
{
#if defined(NPY_HAVE_AVX512_SKX)
    avx512_qselect(arr, kth, num, true);
#elif defined(NPY_HAVE_AVX2)
    avx2_qselect(arr, kth, num, true);
#endif
}
} // anonymous
```
The absence of the `static` keyword or unique names for each target can lead to linking against the lowest CPU feature, irrespective of the number of targets. For example, if this source is compiled for both `avx2` and `avx512_skx`, and the compiler chooses not to inline these functions, then each resulting object will have the same global symbols for `x86_qselect` and `x86_qsort`. During static linking of `x86_simd_qsort_avx2.dispatch.o `and `x86_simd_qsort_avx512_skx.dispatch.o`, the linker will prioritize the first found symbol, and since Meson's features module sorting generated objects by the lowest feature to prevent running unsupported instructions the priority will go to the global symbols of `x86_simd_qsort_avx2.dispatch.o `.

Reply:
Given we build with -O3, compiler should inline these functions and those symbols won't exist. But it does makes sense to play it safe I suppose. That may not be true for debug builds built with -O0.